text/0000-pattern-3.md

```diff
@@ -0,0 +1,1653 @@
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rmatch_indices<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = (H::Target::Index, H)>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn find<H, P>(haystack: H, pattern: P) -> Option<H::Target::Index>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rfind<H, P>(haystack: H, pattern: P) -> Option<H::Target::Index>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn match_ranges<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = (Range<H::Target::Index>, H)>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rmatch_ranges<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = (Range<H::Target::Index>, H)>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn find_range<H, P>(haystack: H, pattern: P) -> Option<Range<H::Target::Index>>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rfind_range<H, P>(haystack: H, pattern: P) -> Option<Range<H::Target::Index>>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+```
+
+**Split**
+
+```rust
+pub fn split<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rsplit<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn split_terminator<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rsplit_terminator<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn splitn<H, P>(haystack: H, n: usize, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rsplitn<H, P>(haystack: H, n: usize, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+```
+
+**Replace**
+
+```rust
+pub fn replace_with<H, P, F, W>(src: H, from: P, replacer: F, writer: W)
+where
+    H: Haystack,
+    P: Pattern<H>,
+    F: FnMut(H) -> H,
+    W: FnMut(H);
+
+pub fn replacen_with<H, P, F, W>(src: H, from: P, replacer: F, n: usize, writer: W)
+where
+    H: Haystack,
+    P: Pattern<H>,
+    F: FnMut(H) -> H,
+    W: FnMut(H);
+```
+
+</details>
+
+Most algorithms are very simple to implement using trisection (`.split_around()`). For instance,
+`split()` can be implemented as:
+
+```rust
+gen fn split<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>,
+{
+    let mut searcher = pattern.into_searcher();
+    let mut rest = Span::from(haystack);
+    while let Some(range) = searcher.search(rest.borrow()) {
+        let [left, _, right] = unsafe { rest.split_around(range) };
```

Comment:
It seems very common to call `split_around` and then throw away one or more of the components. For owned containers like `Vec`, at least, this involves allocating a vector for the ignored elements, copying them to their new location, then finally dropping and deallocating. Would it be possible to add more methods to `Haystack` that only return some of the parts? They could have default definitions in terms of `split_around`, so they shouldn't cause any more difficulty for implementers, but owned containers would be able to override them for better performance.

It also occurs to me that `slice_unchecked` is actually one of these specialized methods, returning only the middle component.

Reply:
We'll need 3 more names for this 😝